File: backend/services/memory/context-compressor.js
// ── Summarizing context compressor ──
// Replaces the oldest middle messages in a conversation with a single fenced
// summary message, while protecting the first `headCount` non-system
// messages and the last `tailCount` non-system messages. Works behind
// session-store.compression_locks so concurrent compactions are safe.
//
// Enable via AUGUST_SUMMARIZING_COMPACTOR=1. Falls back to the legacy
// threshold-based `compactMessagesToThreshold` in backend/adapters/base.js
// when the flag is off or the lock cannot be acquired.

const { estimateTokens } = require('../../lib/tokens');
const sessionStore = require('../storage/session-store');

const DEFAULT_HEAD_COUNT = 4;
const DEFAULT_TAIL_COUNT = 6;
const DEFAULT_SUMMARY_MARKER = '<<compressed_summary';
const DEFAULT_MAX_SUMMARY_CHARS = 2000;
const LOCK_HOLDER = 'summarizing-compactor';
const FEATURE_FLAG = 'AUGUST_SUMMARIZING_COMPACTOR';

function isFeatureEnabled() {
    return process.env[FEATURE_FLAG] === '1';
}

/**
 * Default local summarizer. Joins text content from each middle message,
 * truncates to `maxSummaryChars`, and returns the summary string. When an
 * LLM-based summarizer is wired in, callers pass `options.summarizer`.
 */
function localSummarize(messages, { maxSummaryChars = DEFAULT_MAX_SUMMARY_CHARS } = {}) {
    const lines = [];
    for (const m of messages) {
        const role = m.role || 'unknown';
        let text = '';
        if (typeof m.content === 'string') {
            text = m.content;
        } else if (Array.isArray(m.content)) {
            text = m.content
                .filter(b => b && (b.type === 'text' || b.type === 'output_text'))
                .map(b => b.text || '')
                .join(' ');
        } else if (m.content) {
            try { text = JSON.stringify(m.content); } catch (_) { text = String(m.content); }
        }
        if (m.tool_calls && m.tool_calls.length) {
            const names = m.tool_calls.map(tc => tc.function?.name || tc.name).filter(Boolean);
            if (names.length) text += ` [tool_calls: ${names.join(', ')}]`;
        }
        const trimmed = text.replace(/\s+/g, ' ').trim().slice(0, 600);
        if (trimmed) lines.push(`[${role}] ${trimmed}`);
    }
    let summary = lines.join('\n');
    if (summary.length > maxSummaryChars) {
        summary = summary.slice(0, maxSummaryChars) + '…';
    }
    return summary;
}

/**
 * Build the fenced summary message. Metadata includes the compressed
 * message count and the time so downstream consumers can inspect what was
 * collapsed.
 */
function buildSummaryMessage(middleMessages, summaryText, summaryMarker, at = new Date().toISOString()) {
    const meta = JSON.stringify({
        marker: 'august.summary',
        compressedCount: middleMessages.length,
        at,
    });
    return {
        role: 'system',
        content: `${summaryMarker}\n${meta}\n${summaryText}\n${summaryMarker.replace(/^</, '<\\/')}>>`,
    };
}

/**
 * Compress messages to fit within a token threshold by summarizing the middle
 * while preserving the head and tail. Pure function — does not touch the
 * session store. Caller is responsible for locking and persistence.
 *
 * Returns one of:
 *   { changed: true, messages, summary }   — compaction was applied
 *   { changed: false, messages, reason }  — no compaction needed
 */
function summarizeMessagesToThreshold(messages, tools, threshold, {
    headCount = DEFAULT_HEAD_COUNT,
    tailCount = DEFAULT_TAIL_COUNT,
    summaryMarker = DEFAULT_SUMMARY_MARKER,
    summarizer = localSummarize,
    maxSummaryChars = DEFAULT_MAX_SUMMARY_CHARS,
} = {}) {
    if (!Array.isArray(messages)) return { changed: false, messages: [], reason: 'messages-not-array' };
    if (typeof threshold !== 'number' || threshold <= 0) {
        return { changed: false, messages, reason: 'invalid-threshold' };
    }
    if (estimateTokens(messages, tools) <= threshold) {
        return { changed: false, messages, reason: 'under-threshold' };
    }

    const systemMessages = messages.filter(m => m.role === 'system');
    const otherMessages = messages.filter(m => m.role !== 'system');

    if (otherMessages.length <= headCount + tailCount) {
        return { changed: false, messages, reason: 'too-few-to-summarize' };
    }

    const head = otherMessages.slice(0, headCount);
    const tail = otherMessages.slice(otherMessages.length - tailCount);
    const middle = otherMessages.slice(headCount, otherMessages.length - tailCount);

    // Classify system messages by their position relative to non-system
    // messages. Leading system messages stay at the start; trailing ones stay
    // at the end; the (rare) middle ones go to the start to preserve context.
    const firstNonSystemIdx = messages.findIndex(m => m.role !== 'system');
    const lastNonSystemIdx = messages.length - 1 - [...messages].reverse().findIndex(m => m.role !== 'system');
    const leadingSystem = [];
    const trailingSystem = [];
    for (const sm of systemMessages) {
        const originalIdx = messages.indexOf(sm);
        if (originalIdx < firstNonSystemIdx) leadingSystem.push(sm);
        else if (originalIdx > lastNonSystemIdx) trailingSystem.push(sm);
        else leadingSystem.push(sm);
    }

    const summaryText = summarizer(middle, { maxSummaryChars });
    const summaryMessage = buildSummaryMessage(middle, summaryText, summaryMarker);

    const rebuilt = [...leadingSystem, ...head, summaryMessage, ...tail, ...trailingSystem];
    const rebuiltTokens = estimateTokens(rebuilt, tools);

    return {
        changed: true,
        messages: rebuilt,
        summary: {
            headCount,
            tailCount,
            compressedCount: middle.length,
            originalTokens: estimateTokens(messages, tools),
            compressedTokens: rebuiltTokens,
            underThreshold: rebuiltTokens <= threshold,
        },
    };
}

/**
 * Acquire the compression lock, run `summarizeMessagesToThreshold`, and
 * release the lock. Returns the compaction result (or null if the lock is
 * held by another holder).
 *
 * Pure helper that does not write to the session store; callers decide
 * whether to persist the rebuilt messages via `replaceMessages` or
 * `rewindToMessage`.
 */
async function compactWithLock(sessionId, messages, tools, threshold, options = {}) {
    if (!sessionId || !sessionStore.isReady()) return null;
    const acquired = sessionStore.acquireCompressionLock(sessionId, LOCK_HOLDER, 300);
    if (!acquired) return { changed: false, messages, reason: 'lock-held' };
    try {
        return summarizeMessagesToThreshold(messages, tools, threshold, options);
    } finally {
        sessionStore.releaseCompressionLock(sessionId);
    }
}

module.exports = {
    summarizeMessagesToThreshold,
    localSummarize,
    buildSummaryMessage,
    compactWithLock,
    isFeatureEnabled,
    FEATURE_FLAG,
    LOCK_HOLDER,
    DEFAULT_HEAD_COUNT,
    DEFAULT_TAIL_COUNT,
    DEFAULT_SUMMARY_MARKER,
    DEFAULT_MAX_SUMMARY_CHARS,
};

File: backend-py/app/services/memory/context_compressor.py
"""
Context compressor — summarizes middle messages to fit within token thresholds
while preserving head and tail messages.

Port of backend/services/memory/context-compressor.js (177 lines).
"""

from __future__ import annotations
import os
from typing import Callable
from app.jsonUtils import as_list, as_str
from app.providers.clients.base import estimateTokens

DEFAULT_HEAD_COUNT = 4
DEFAULT_TAIL_COUNT = 6
DEFAULT_SUMMARY_MARKER = '<<compressed_summary'
DEFAULT_MAX_SUMMARY_CHARS = 2000
FEATURE_FLAG = 'AUGUST_SUMMARIZING_COMPACTOR'


def isFeatureEnabled() -> bool:
    """Check if the summarizing compactor feature flag is set.

    Enabled by default — the env var AUGUST_SUMMARIZING_COMPACTOR
    can be set to "0" to disable."""
    val = os.environ.get(FEATURE_FLAG)
    if val is not None:
        return val == '1'
    return True


def localSummarize(messages: list[dict[str, object]], maxSummaryChars: int = DEFAULT_MAX_SUMMARY_CHARS) -> str:
    """Default local summarizer.

    Joins text content from each message, truncates to max_summary_chars,
    and returns the summary string.
    """
    lines: list[str] = []
    for msg in messages:
        role = msg.get('role', 'unknown')
        content = msg.get('content', '')
        text = ''
        if isinstance(content, str):
            text = content
        elif isinstance(content, list):
            text = ' '.join(
                (
                    str(b.get('text', ''))
                    for b in content
                    if isinstance(b, dict) and b.get('type') in ('text', 'output_text')
                )
            )
        elif content:
            try:
                import json

                text = json.dumps(content)
            except (TypeError, ValueError):
                text = str(content)
        toolCalls = as_list(msg.get('tool_calls'))
        if toolCalls:
            names: list[str] = []
            for tc in toolCalls:
                if not isinstance(tc, dict):
                    continue
                func = tc.get('function')
                name = as_str(func.get('name')) if isinstance(func, dict) else ''
                if not name:
                    name = as_str(tc.get('name'))
                if name:
                    names.append(name)
            if names:
                text += f' [tool_calls: {", ".join(names)}]'
        trimmed = ' '.join(text.split())[:600]
        if trimmed:
            lines.append(f'[{role}] {trimmed}')
    summary = '\n'.join(lines)
    if len(summary) > maxSummaryChars:
        summary = summary[:maxSummaryChars] + '…'
    return summary


def buildSummaryMessage(
    middleMessages: list[dict[str, object]], summaryText: str, summaryMarker: str = DEFAULT_SUMMARY_MARKER
) -> dict[str, object]:
    """Build a fenced summary message from the middle messages."""
    import json

    meta = json.dumps({'marker': 'august.summary', 'compressed_count': len(middleMessages)})
    return {
        'role': 'system',
        'content': f'{summaryMarker}\n{meta}\n{summaryText}\n{summaryMarker.replace("<", "</")}>>',
    }


def _isSummaryMessage(msg: dict[str, object], summaryMarker: str = DEFAULT_SUMMARY_MARKER) -> bool:
    """True if msg is a prior compressed-summary system block.

    Detected by the opening marker (`<<compressed_summary`) in a system
    message's string content. The compactor emits exactly this shape from
    build_summary_message, so this reliably identifies prior summaries that
    would otherwise accumulate across repeated compactions (s4).
    """
    if msg.get('role') != 'system':
        return False
    content = msg.get('content', '')
    if not isinstance(content, str):
        return False
    return content.startswith(summaryMarker)


def _extractSummaryText(msg: dict[str, object], summaryMarker: str = DEFAULT_SUMMARY_MARKER) -> str:
    """Recover the human summary text from a fenced summary message.

        build_summary_message emits
        ``{marker}
    {meta_json}
    {summary_text}
    {closing}``. Drop the first line
        (marker), the second (meta json), and the last (closing marker) to get the
        body. Returns "" if the shape is unexpected.
    """
    content = msg.get('content', '')
    if not isinstance(content, str) or not content.startswith(summaryMarker):
        return ''
    lines = content.split('\n')
    if len(lines) < 3:
        return ''
    return '\n'.join(lines[2:-1])


def compressMessages(
    messages: list[dict[str, object]],
    threshold: int,
    head_count: int = DEFAULT_HEAD_COUNT,
    tail_count: int = DEFAULT_TAIL_COUNT,
    summarizer: Callable | None = None,
) -> list[dict[str, object]]:
    """Compress messages to fit within a token threshold by summarizing the middle.

    Preserves the first ``head_count`` and last ``tail_count`` messages,
    summarizing everything in between.

    Args:
        messages: Full conversation messages.
        threshold: Token threshold to compress under.
        head_count: Number of messages to preserve at the start.
        tail_count: Number of messages to preserve at the end.
        summarizer: Optional async callable that returns a summary string.

    Returns:
        Compressed message list (may be unchanged if already under threshold).
    """
    if not messages:
        return messages
    currentTokens = estimateTokens(messages)
    if currentTokens <= threshold:
        return list(messages)
    nonSystem = [m for m in messages if m.get('role') != 'system']
    systemMsgs = [m for m in messages if m.get('role') == 'system']
    priorSummaryTexts = [_extractSummaryText(m) for m in systemMsgs if _isSummaryMessage(m)]
    priorSummaryTexts = [t for t in priorSummaryTexts if t]
    otherSystem = [m for m in systemMsgs if not _isSummaryMessage(m)]
    if len(nonSystem) <= head_count + tail_count:
        return list(messages)
    head = nonSystem[:head_count]
    tail = nonSystem[-tail_count:]
    middle = nonSystem[head_count:-tail_count] if tail_count > 0 else nonSystem[head_count:]
    if not middle:
        return list(messages)
    if summarizer:
        summaryText = summarizer(middle)
    else:
        summaryText = localSummarize(middle)
    if priorSummaryTexts:
        summaryText = 'Earlier summary:\n' + '\n---\n'.join(priorSummaryTexts) + '\n\nRecent summary:\n' + summaryText
    summaryMsg = buildSummaryMessage(middle, summaryText)
    compressed = otherSystem + head + [summaryMsg] + tail
    compressedTokens = estimateTokens(compressed)
    if compressedTokens >= currentTokens:
        return list(messages)
    return compressed
